### PR TITLE
enh (service): Create new API endpoint to create a service template

### PR DESCRIPTION
**Pull request by callapa**
## Description

Create a new API endpoint to create a service template
```
POST /configuration/services/templates
```

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    type: integer
+    description: |
+      Define the number of 'time units' between regularly scheduled checks of the service.
+      
+      With the default time unit of 60s, this number will mean multiples of 1 minute.
+    nullable: true
+  retry_check_interval:
+    type: integer
+    description: |
+      Define the number of "time units" to wait before scheduling a re-check for this service after a non-UP state was detected.
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:53]:
```suggestion
      Define the number of "time units" to wait before scheduling a re-check for this service after a non-OK state was detected.
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      it will revert to being scheduled at its "normal" check interval rate.
+    nullable: true
+  active_check_enabled:
+    type: integer
+    description: |
+      Indicates whether active checks are enabled or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:67]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    example: 2
+  passive_check_enabled:
+    type: integer
+    description: |
+      Indicates whether passive checks are enabled or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:76]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    example: 2
+  volatility_enabled:
+    type: integer
+    description: |
+      Indicates whether the service is "volatile" or not
+
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:85]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    example: 2
+  notification_enabled:
+    type: integer
+    description: |
+      Specify whether notifications are enabled or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:94]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      The value is the sum of all the values of the selected options.
+      
+      * `1` - WARNING
+      * `2` - UNKNOWN
+      * `4` - CRITICAL
+      * `8` - RECOVERY
+      * `16` - FLAPPING
+      * `32` - DOWNTIME_SCHEDULED
+      
+      example: A value equal to 5 corresponding to the selected options WARNING and CRITICAL
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:137]:
```suggestion
      Example: A value equal to 5 corresponds to the selected options WARNING and CRITICAL
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    example: 5
+  first_notification_delay:
+    type: integer
+    nullable: true
+    description: |
+      Define the number of "time units" to wait before sending out the first problem notification when this service enters a non-UP state.
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:143]:
```suggestion
      Define the number of "time units" to wait before sending out the first problem notification when this service enters a non-OK state.
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+      
+      With the default time unit of 60s, this number will mean multiples of 1 minute.
+  recovery_notification_delay:
+    type: integer
+    nullable: true
+    description: |
+      Define the number of "time units" to wait before sending out the recovery notification when this service enters an UP state.
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:150]:
```suggestion
      Define the number of "time units" to wait before sending out the recovery notification when this service enters an OK state.
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    description: "Specify a duration of acknowledgement for this service template."
+    nullable: true
+  freshness_checked:
+    type: integer
+    description: |
+      Indicates whether freshness is checked or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:164]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    description: "Specify the freshness threshold (in seconds) for this service."
+    nullable: true
+  flap_detection_enabled:
+    type: integer
+    description: |
+      Indicates whether the flap detection is enabled or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:177]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    description: "Specify the high state change threshold used in flap detection for this service."
+    nullable: true
+  event_handler_enabled:
+    type: integer
+    description: |
+      Indicates whether the event handler is enabled or not
+      
+      * `0` - STATUS_DISABLED
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of the Centreon Engine)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:194]:
```suggestion
      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
```

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    nullable: true
+    maxLength: 65535
+  note_url:
+    type: string
+    description: "Define an optional URL that can be used to provide more information about the service."
+    maxLength: 65535
+    nullable: true
+  action_url:
+    type: string
+    description: "Define an optional URL that can be used to provide more actions to be performed on the service."
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/Schema/NewServiceTemplate.yaml:226]:
```suggestion
    description: "Define an optional URL that can be used to specify actions to be performed on the service."
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+
+    ($this->useCase)(new AddServiceTemplateRequest(), $this->presenter);
+
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ForbiddenResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::addNotAllowed()->getMessage());
+});
+
+it('should present a ErrorResponse when the user name already exists', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:121]:
```suggestion
it('should present an ErrorResponse when the user name already exists', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+    $request = new AddServiceTemplateRequest();
+    $request->name = $name;
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ErrorResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::nameAlreadyExists($name)->getMessage());
+});
+
+it('should present a ConflictResponse when the severity id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:148]:
```suggestion
it('should present a ConflictResponse when the severity ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+        ->willReturn(false);
+
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ConflictResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::idDoesNotExist('severity_id', $request->severityId)->getMessage());
+});
+
+it('should present a ConflictResponse when the performance graph id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:181]:
```suggestion
it('should present a ConflictResponse when the performance graph ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 16]
```diff
@@ -0,0 +1,714 @@
...
+        ->willReturn(false);
+
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ConflictResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::idDoesNotExist('graph_template_id', $request->severityId)->getMessage());
+});
+
+it('should present a ConflictResponse when the service template id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:221]:
```suggestion
it('should present a ConflictResponse when the service template ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 17]
```diff
@@ -0,0 +1,714 @@
...
+        ->willReturn(false);
+
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ConflictResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::idDoesNotExist('service_template_id', $request->severityId)->getMessage());
+});
+
+it('should present a ConflictResponse when the command id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:268]:
```suggestion
it('should present a ConflictResponse when the command ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 18]
```diff
@@ -0,0 +1,714 @@
...
+        ->willReturn(false);
+
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ConflictResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(ServiceTemplateException::idDoesNotExist('check_command_id', $request->severityId)->getMessage());
+});
+
+it('should present a ConflictResponse when the event handler id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:322]:
```suggestion
it('should present a ConflictResponse when the event handler ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 19]
```diff
@@ -0,0 +1,714 @@
...
+        ->and($this->presenter->response->getMessage())
+        ->toBe(
+            ServiceTemplateException::idDoesNotExist(
+                'event_handler_command_id',
+                $request->eventHandlerId
+            )->getMessage()
+        );
+});
+
+it('should present a ConflictResponse when the time period id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:386]:
```suggestion
it('should present a ConflictResponse when the time period ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 20]
```diff
@@ -0,0 +1,714 @@
...
+        ->and($this->presenter->response->getMessage())
+        ->toBe(
+            ServiceTemplateException::idDoesNotExist(
+                'check_timeperiod_id',
+                $request->checkTimePeriodId
+            )->getMessage()
+        );
+});
+
+it('should present a ConflictResponse when the notification time period id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:457]:
```suggestion
it('should present a ConflictResponse when the notification time period ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 21]
```diff
@@ -0,0 +1,714 @@
...
+        ->and($this->presenter->response->getMessage())
+        ->toBe(
+            ServiceTemplateException::idDoesNotExist(
+                'notification_timeperiod_id',
+                $request->notificationTimePeriodId
+            )->getMessage()
+        );
+});
+
+it('should present a ConflictResponse when the icon id is not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:533]:
```suggestion
it('should present a ConflictResponse when the icon ID is not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 22]
```diff
@@ -0,0 +1,714 @@
...
+        ->and($this->presenter->response->getMessage())
+        ->toBe(
+            ServiceTemplateException::idDoesNotExist(
+                'icon_id',
+                $request->iconId
+            )->getMessage()
+        );
+});
+
+it('should present a InvalidArgumentResponse when data are not valid', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:616]:
```suggestion
it('should present an InvalidArgumentResponse when data are not valid', function () {
```

File: centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php [hunk 23]
```diff
@@ -0,0 +1,714 @@
...
+        ->with($request->iconId)
+        ->willReturn(true);
+
+    ($this->useCase)($request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(InvalidArgumentResponse::class)
+        ->and($this->presenter->response->getMessage());
+});
+
+it('should present a ErrorResponse when an exception is thrown', function () {
```
Comment (cg-tw) [inline centreon/tests/php/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateTest.php:693]:
```suggestion
it('should present an ErrorResponse when an exception is thrown', function () {
```

**Review by cg-tw — APPROVED** [review]

File: centreon/doc/API/v23.10/Configuration/ServiceTemplate/AddOneAndFindServiceTemplates.yaml [hunk 24]
```diff
@@ -35,3 +35,26 @@ get:
...
+      allOf:
+        - $ref: '../../Common/Response/Created.yaml'
+        - content:
+            application/json:
+              schema:
+                $ref: 'Schema/ServiceTemplate.yaml'
+    '403':
+      $ref: '../../Common/Response/Forbidden.yaml'
+    '500':
+      $ref: '../../Common/Response/InternalServerError.yaml'
```
Comment (a-launois) [inline centreon/doc/API/v23.10/Configuration/ServiceTemplate/AddOneAndFindServiceTemplates.yaml:60]:
```suggestion
    '400':
      $ref: '../../Common/Response/BadRequest.yaml'
    '403':
      $ref: '../../Common/Response/Forbidden.yaml'
    '409':
      $ref: '../../Common/Response/Conflict.yaml'
    '500':
      $ref: '../../Common/Response/InternalServerError.yaml'
```

File: centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateRequest.php [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+
+    public string $alias = '';
+
+    /** @var list<string> */
+    public array $commandArguments = [];
+
+    /** @var list<string> */
+    public array $eventHandlerArguments = [];
+
+    public int $notificationTypes = 0;
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplateRequest.php:38]: cf. previous comments
```suggestion
    public int|null $notificationTypes = null;
```

File: centreon/src/Core/ServiceTemplate/Domain/Model/ServiceTemplate.php [hunk 26]
```diff
@@ -139,21 +138,32 @@ public function __construct(
...
                 'actionUrl' => self::MAX_ACTION_URL_LENGTH,
                 'iconAlternativeText' => self::MAX_ICON_ALT_LENGTH,
             ] as $field => $limitation
         ) {
-            $propertyValue = ${$field};
+            $propertyValue = $this->{$field};
+            if ($field === 'alias') {
+                $propertyValue = preg_replace('/\s{2,}/', ' ', $propertyValue);
+                if ($propertyValue === null) {
+                    throw AssertionException::notNull($className . '::alias');
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Domain/Model/ServiceTemplate.php:151]: Relates to comments on NewServiceTemplate
```suggestion
            if ($field === 'name') {
                $propertyValue = preg_replace('/\s{2,}/', ' ', $propertyValue);
                if ($propertyValue === null) {
                    throw AssertionException::notNull($className . '::name');
```

File: centreon/src/Core/ServiceTemplate/Domain/Model/ServiceTemplate.php [hunk 27]
```diff
@@ -139,21 +138,32 @@ public function __construct(
...
                 $this->{$field} = trim($propertyValue);
                 Assertion::notEmptyString($this->{$field}, "{$className}::{$field}");
                 Assertion::maxLength($this->{$field}, $limitation, "{$className}::{$field}");
             }
         }
 
+        Assertion::unauthorisedCharacters(
+            $this->alias,
+            MonitoringServer::ILLEGAL_CHARACTERS,
+            "{$className}::alias"
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Domain/Model/ServiceTemplate.php:164]: Relates to comments on NewServiceTemplate
```suggestion
            $this->name,
            MonitoringServer::ILLEGAL_CHARACTERS,
            "{$className}::name"
```

File: centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php [hunk 28]
```diff
@@ -0,0 +1,911 @@
...
+            throw AssertionException::notNull($this->className . '::alias');
+        }
+        $alias = trim($alias);
+        Assertion::notEmptyString($alias, $this->className . '::alias');
+        Assertion::maxLength($alias, self::MAX_ALIAS_LENGTH, $this->className . '::alias');
+        Assertion::unauthorisedCharacters(
+            $alias,
+            MonitoringServer::ILLEGAL_CHARACTERS,
+            $this->className . '::alias'
+        );
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php:179]: Same as previous, this control is applied on name property not alias on UI

File: centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplate.php [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+        foreach (NotificationTypeConverter::fromBits($request->notificationTypes) as $notificationType) {
+            $serviceTemplate->addNotificationType($notificationType);
+        }
+
+        $serviceTemplate->setContactAdditiveInheritance(
+            $inheritanceMode ? $request->isContactAdditiveInheritance : false
+        );
+
+        $serviceTemplate->setContactGroupAdditiveInheritance(
+            $inheritanceMode ? $request->isContactGroupAdditiveInheritance : false
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplate.php:253]: $inheritanceMode could be egal to '1' , '2' or '3'
```suggestion
            $inheritanceMode === '1' ? $request->isContactAdditiveInheritance : false
        );

        $serviceTemplate->setContactGroupAdditiveInheritance(
            $inheritanceMode === '1' ? $request->isContactGroupAdditiveInheritance : false
```

File: centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateOnPremSchema.json [hunk 30]
```diff
@@ -0,0 +1,239 @@
...
+             "type": [
+                "null",
+                "integer"
+            ]
+        },
+        "is_activated": {
+            "type": "boolean"
+        },
+        "is_locked": {
+            "type": "boolean"
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateOnPremSchema.json:236]: Relates to comments on doc API, creation of service template with is_locked = true is not accepted outside of plugin/monitoring connectors installation (to confirm with @omercier if needed)
```suggestion
```

File: centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateSaasSchema.json [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+            ]
+        },
+        "severity_id": {
+             "type": [
+                "null",
+                "integer"
+            ]
+        },
+        "is_locked": {
+            "type": "boolean"
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateSaasSchema.json:54]: Same as above
```suggestion
```

File: centreon/src/Core/ServiceTemplate/Infrastructure/Model/NotificationTypeConverter.php [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * @param NotificationType[] $notificationTypes
+     *
+     * @return int
+     */
+    public static function toBits(array $notificationTypes): int
+    {
+        if ($notificationTypes === []) {
+            return self::ALL_TYPE;
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/Model/NotificationTypeConverter.php:54]: Empty array mean, values are inherited and therefore could be different than self::ALL_TYPE

```suggestion
    {
        if ($notificationTypes === []) {
            return null;
```

File: centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php [hunk 33]
```diff
@@ -40,6 +43,238 @@ public function __construct(DatabaseConnection $db)
...
+        );
+        $statement->bindValue(
+            ':contact_additive_inheritance',
+            $newServiceTemplate->isContactAdditiveInheritance(),
+            \PDO::PARAM_INT
+        );
+        $statement->bindValue(':command_id', $newServiceTemplate->getCommandId(), \PDO::PARAM_INT);
+        $statement->bindValue(
+            ':command_arguments',
+            $this->serializeArguments($newServiceTemplate->getCommandArguments())
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php:136]: What about str_replace(["\n", "\t", "\r"], ['#BR#', '#T#', '#R#'], $argument) ?

File: centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php [hunk 34]
```diff
@@ -40,6 +43,238 @@ public function __construct(DatabaseConnection $db)
...
+        );
+        $statement->bindValue(':event_handler_id', $newServiceTemplate->getEventHandlerId(), \PDO::PARAM_INT);
+        $statement->bindValue(
+            ':event_handler_arguments',
+            $this->serializeArguments($newServiceTemplate->getEventHandlerArguments())
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php:141]: Same question as above

File: centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php [hunk 35]
```diff
@@ -40,6 +43,238 @@ public function __construct(DatabaseConnection $db)
...
+            $newServiceTemplate->getMaxCheckAttempts(),
+            \PDO::PARAM_INT
+        );
+        $statement->bindValue(
+            ':description',
+            $newServiceTemplate->getName()
+        );
+        $statement->bindValue(
+            ':comment',
+            $newServiceTemplate->getComment()
```
Comment (a-launois) [inline centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php:206]: No use of repositoryTrait->legacyHtmlEncode for string parameters ?

**Review by a-launois — CHANGES_REQUESTED** [review]

File: centreon/tests/php/Centreon/Domain/Common/AssertionTest.php [hunk 36]
```diff
@@ -309,6 +314,9 @@ function () use ($propertyPath, $failAssertion, $failMessage, $failCode): void {
     "isInstanceOf(new \Exception(), \Throwable::class)" => [
         fn() => Assertion::isInstanceOf(new \Exception(), \Throwable::class, $propertyPath),
     ],
+    "unauthorisedCharacters('new bad string!', '&#@')" => [
```
Comment (abuathier) [inline centreon/tests/php/Centreon/Domain/Common/AssertionTest.php:317]: Missing tests with empty string `''`

File: centreon/src/Core/PerformanceGraph/Infrastructure/Repository/DbReadPerformanceGraphRepository.php [hunk 37]
```diff
@@ -0,0 +1,48 @@
...
+        $this->db = $db;
+    }
+
+    public function exists(int $performanceGraphId): bool
+    {
+        $request = $this->translateDbName('SELECT * FROM `:db`.giv_graphs_template WHERE graph_id = :id');
+        $statement = $this->db->prepare($request);
+        $statement->bindValue(':id', $performanceGraphId, \PDO::PARAM_INT);
+        $statement->execute();
+        return (bool) $statement->fetchColumn();
```
Comment (abuathier) [inline centreon/src/Core/PerformanceGraph/Infrastructure/Repository/DbReadPerformanceGraphRepository.php:46]: php-cs-fixer adds a line feed before `return`
```suggestion

        return (bool) $statement->fetchColumn();
```

File: centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplate.php [hunk 38]
```diff
@@ -0,0 +1,373 @@
...
+
+    /**
+     * @param AddServiceTemplateRequest $request
+     *
+     * @throws \Exception
+     * @throws AssertionFailedException
+     *
+     * @return NewServiceTemplate
+     */
+    private function createNewServiceTemplate(AddServiceTemplateRequest $request): NewServiceTemplate
```
Comment (abuathier) [inline centreon/src/Core/ServiceTemplate/Application/UseCase/AddServiceTemplate/AddServiceTemplate.php:228]: Maybe move the create internal code (100+ lines) inside a Factory ?

File: centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php [hunk 39]
```diff
@@ -0,0 +1,911 @@
...
+
+    private ?int $highFlapThreshold;
+
+    private ?int $notificationInterval;
+
+    private ?int $recoveryNotificationDelay;
+
+    private ?int $firstNotificationDelay;
+
+    private ?int $acknowledgementTimeout;
```
Comment (abuathier) [inline centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php:123]:
```suggestion
    private ?string $comment = null;

    private ?string $note = null;

    private ?string $noteUrl = null;

    private ?string $actionUrl = null;

    private ?string $iconAlternativeText = null;

    private ?int $graphTemplateId = null;

    private ?int $serviceTemplateParentId = null;

    private ?int $commandId = null;

    private ?int $eventHandlerId = null;

    private ?int $notificationTimePeriodId = null;

    private ?int $checkTimePeriodId = null;

    private ?int $iconId = null;

    private ?int $severityId = null;

    private ?int $maxCheckAttempts = null;

    private ?int $normalCheckInterval = null;

    private ?int $retryCheckInterval = null;

    private ?int $freshnessThreshold = null;

    private ?int $lowFlapThreshold = null;

    private ?int $highFlapThreshold = null;

    private ?int $notificationInterval = null;

    private ?int $recoveryNotificationDelay = null;

    private ?int $firstNotificationDelay = null;

    private ?int $acknowledgementTimeout = null;
```

File: centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateController.php [hunk 40]
```diff
@@ -0,0 +1,180 @@
...
+        $this->denyAccessUnlessGrantedForApiConfiguration();
+
+        $presenter = $isCloudPlatform ? $saasPresenter : $onPremPresenter;
+        $validationSchema = $isCloudPlatform
+            ? 'AddServiceTemplateSaasSchema.json'
+            : 'AddServiceTemplateOnPremSchema.json';
+
+        try {
+            /** @var _ServiceTemplate $data
+             */
```
Comment (abuathier) [inline centreon/src/Core/ServiceTemplate/Infrastructure/API/AddServiceTemplate/AddServiceTemplateController.php:113]:
```suggestion
            /** @var _ServiceTemplate $data */
```

File: centreon/src/Centreon/Domain/Common/Assertion/AssertionException.php [hunk 41]
```diff
@@ -47,6 +46,9 @@ class AssertionException extends \Assert\InvalidArgumentException
     public const VALUE_EMPTY = Assert::VALUE_EMPTY;
     public const VALUE_NULL = Assert::VALUE_NULL;
 
+    // Error codes of Centreon assertion start from 300
+    public const INVALID_CHARACTERS = 300;
```
[the remark this answers is not in the document]
Comment (callapa) [reply centreon/src/Centreon/Domain/Common/Assertion/AssertionException.php:50]: @abuathier Where ?

File: centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php (hunk 35 above)
Comment (callapa) [reply centreon/src/Core/ServiceTemplate/Infrastructure/Repository/DbWriteServiceTemplateRepository.php:206]: No, comments are not encoded in database.

File: centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php [hunk 42]
```diff
@@ -0,0 +1,911 @@
...
+     * @param string $alias
+     *
+     * @throws AssertionFailedException
+     */
+    public function setAlias(string $alias): void
+    {
+        $alias = preg_replace('/\s{2,}/', ' ', $alias);
+        if ($alias === null) {
+            throw AssertionException::notNull($this->className . '::alias');
+        }
```
[the remark this answers is not in the document]
Comment (callapa) [reply centreon/src/Core/ServiceTemplate/Domain/Model/NewServiceTemplate.php:171]: Yes for both

**Review by abuathier — APPROVED** [review]

**Review by a-launois — DISMISSED** [review]

**Review by abuathier — APPROVED** [review]

**Review by cg-tw — APPROVED** [review]
